packages/react-native-reanimated/Common/cpp/LayoutAnimations/LayoutAnimationsUtils.h

```diff
@@ -0,0 +1,174 @@
+struct Frame {
+  std::optional<double> x, y, width, height;
+  Frame(jsi::Runtime &runtime, const jsi::Object &newStyle) {
+    if (newStyle.hasProperty(runtime, "originX")) {
+      x = newStyle.getProperty(runtime, "originX").asNumber();
+    }
+    if (newStyle.hasProperty(runtime, "originY")) {
+      y = newStyle.getProperty(runtime, "originY").asNumber();
+    }
+    if (newStyle.hasProperty(runtime, "width")) {
+      width = newStyle.getProperty(runtime, "width").asNumber();
+    }
+    if (newStyle.hasProperty(runtime, "height")) {
+      height = newStyle.getProperty(runtime, "height").asNumber();
+    }
+  }
+};
+
+struct UpdateValues {
+  Props::Shared newProps;
+  Frame frame;
+};
+
+struct Snapshot {
+  double x, y, width, height, windowWidth, windowHeight;
+  Snapshot(const ShadowView &shadowView, Window window) {
+    auto &frame = shadowView.layoutMetrics.frame;
+    x = frame.origin.x;
+    y = frame.origin.y;
+    width = frame.size.width;
+    height = frame.size.height;
+    windowWidth = window.width;
+    windowHeight = window.height;
+  }
+};
+
+typedef enum ExitingState {
+  UNDEFINED = 1,
+  WAITING = 2,
+  ANIMATING = 4,
+  DEAD = 8,
+  MOVED = 16,
+  DELETED = 32,
+} ExitingState;
+
+struct MutationNode;
+
+/**
+ Represents a view that was either removed or had a child removed from the
+ ShadowTree
+ */
+struct Node {
+  std::vector<std::shared_ptr<MutationNode>> children, unflattenedChildren;
+  std::shared_ptr<Node> parent, unflattenedParent;
+  Tag tag;
+  void removeChild(std::shared_ptr<MutationNode> child);
+  void handleMutation(ShadowViewMutation mutation);
+  void insertChildren(std::vector<std::shared_ptr<MutationNode>> &newChildren);
+  void insertUnflattenedChildren(std::vector<std::shared_ptr<MutationNode>> &newChildren);
+  explicit Node(Tag tag) : tag(tag) {}
+  Node(Node &&node) : children(std::move(node.children)), unflattenedChildren(std::move(node.unflattenedChildren)), tag(node.tag) {}
+};
+
+/**
+ Represents a view that was removed from the ShadowTree
+ */
+struct MutationNode : public Node {
+  ShadowViewMutation mutation;
+  std::unordered_set<Tag> animatedChildren;
+  ExitingState state = UNDEFINED;
+  explicit MutationNode(ShadowViewMutation &mutation)
+      : Node(mutation.oldChildShadowView.tag), mutation(mutation) {}
+  MutationNode(ShadowViewMutation &mutation, Node &&node)
+      : Node(std::move(node)), mutation(mutation) {}
+};
+
+struct SurfaceManager {
+  mutable std::unordered_map<
+      SurfaceId,
+      std::shared_ptr<std::unordered_map<Tag, UpdateValues>>>
+      props_;
+  mutable std::unordered_map<SurfaceId, Window> windows_;
+
+  std::unordered_map<Tag, UpdateValues> &getUpdateMap(SurfaceId surfaceId) {
+    auto props = props_.find(surfaceId);
+    if (props != props_.end()) {
+      return *props->second;
+    }
+
+    auto newProps = std::make_shared<std::unordered_map<Tag, UpdateValues>>();
+    props_.insert_or_assign(surfaceId, newProps);
+    return *newProps;
+  }
```

Comment:
The `newProps` shared pointer will get deallocated. But since I put a copy of this shared pointer in the `props_` map, the underlying object won't be deallocated. So `*newProps` returns a reference to the underlying `unordered_map`, which is what I want. Am I missing something?

Reply:
Right 👍